### DISCUSSION TITLE
Expected match not being found

### DISCUSSION THREAD

**Discussion by reyvasquez-vh** [moj-analytical-services/splink#1139] [Q&A] [answered]
I'm trying to cluster data of a single file--not dedupe and remove duplicates, but just identify them.
The way I had approached this is to perform a "link_only" on two identical files.  Most matches are detected, but there are a few records that are not being matched (compared), and I'm trying to figure out what I'm doing wrong.

Below are my settings, blocking, and training parameters:

```
settings = {
    "probability_two_random_records_match": 1 / 100000,
    "link_type": "link_only",
    "comparisons": [
        cl.levenshtein_at_thresholds("first", distance_threshold_or_thresholds=[1, 2]),
        cl.levenshtein_at_thresholds("last", distance_threshold_or_thresholds=[1, 2]),
        cl.exact_match("dob"),
        cl.exact_match("ssn"),
        cl.levenshtein_at_thresholds("address1", distance_threshold_or_thresholds=[1,2]),
    ],
    "blocking_rules_to_generate_predictions": [
        "substr(l.first, 1, 2) = substr(r.first, 1, 2) and l.last = r.last and l.dob = r.dob",
        "substr(l.first, 1, 2) = substr(r.first, 1, 2) and l.last = r.last and l.ssn = r.ssn",
        "l.dob = r.dob and l.ssn = r.ssn",
    ],
    "retain_matching_columns": True,
    "retain_intermediate_calculation_columns": True,
}
```

```
training_blocking_rule = """substr(l.first, 1, 2) = substr(r.first, 1, 2) and
                            substr(l.last, 1, 2) = substr(r.last, 1, 2) and
                            l.ssn = r.ssn"""
linker.count_num_comparisons_from_blocking_rule(training_blocking_rule)


training_blocking_rule = """substr(l.first, 1, 2) = substr(r.first, 1, 2) and
                            substr(l.last, 1, 2) = substr(r.last, 1, 2) and
                            l.dob = r.dob"""
linker.count_num_comparisons_from_blocking_rule(training_blocking_rule)


training_blocking_rule = """l.dob = r.dob and l.ssn = r.ssn"""
linker.count_num_comparisons_from_blocking_rule(training_blocking_rule)

df_predictions = linker.predict(threshold_match_probability=0.9, threshold_match_weight=5)

clusters = linker.cluster_pairwise_predictions_at_threshold(
            df_predictions, threshold_match_probability=0.9)
```

Partial result below.  My expectation is that ALL FOUR RECORDS should have been clustered together.  Instead, only the corresponding exact records (see unique_id) are being matched, even though all four have identical first, last, dob, and ssn.
My guess is no pair match was generated for **unique_id** `e1f29081eb684274aefbbf1e9dacd669` and `96380c6501384d7eb833e7af2513a80f`.  But based on the blocking above there should have been.

```
+---------------------------------------------+--------------+-----+------+----------+-----------+-----+-----------------+----------+-------------------+--------------------------------+
|cluster_id                                   |source_dataset|first|last  |dob       |ssn        |mrn  |email            |phone1    |address1           |unique_id                       |
+---------------------------------------------+--------------+-----+------+----------+-----------+-----+-----------------+----------+-------------------+--------------------------------+
|left_data-__-e1f29081eb684274aefbbf1e9dacd669|left_data     |john |doedoe|1950-07-04|123-45-6789|null |Johndoe@yahoo.com|5551234567|1234 12rd st nw    |e1f29081eb684274aefbbf1e9dacd669|
|left_data-__-96380c6501384d7eb833e7af2513a80f|left_data     |john |doedoe|1950-07-04|123-45-6789|12345|JOHNDOE@YAHOO.COM|5557654321|1234 12rd street nw|96380c6501384d7eb833e7af2513a80f|
|left_data-__-96380c6501384d7eb833e7af2513a80f|right_data    |john |doedoe|1950-07-04|123-45-6789|12345|JOHNDOE@YAHOO.COM|5557654321|1234 12rd street nw|96380c6501384d7eb833e7af2513a80f|
|left_data-__-e1f29081eb684274aefbbf1e9dacd669|right_data    |john |doedoe|1950-07-04|123-45-6789|null |Johndoe@yahoo.com|5551234567|1234 12rd st nw    |e1f29081eb684274aefbbf1e9dacd669|
+---------------------------------------------+--------------+-----+------+----------+-----------+-----+-----------------+----------+-------------------+--------------------------------+
```

Any help is appreciated.  Thanks!

**Comment by ADBond** [comment, marked as answer]
You are correct that this record pair should be generated by the blocking rules, and so it should be scored. My guess would be that the `match_probability` between these records is too low for the threshold you set, and so is not showing up in your clusters (and is in fact filtered out at the predict step - a `threshold_match_weight` of 5 corresponds to a `threshold_match_probability` of ~0.97, and so you are already filtering out record pairs scoring below this threshold).

I'm not sure if you have accidentally missed it out of the code pasted here, but you do not appear to have any steps to estimate the model parameters, which could explain why you are getting unexpected results (as the predict step will just be running with default Splink values, which may not be suitable for your case). If that is indeed the case you will need to run some estimation steps - see [this section of the docs](https://moj-analytical-services.github.io/splink/demos/04_Estimating_model_parameters.html#estimate-the-parameters-of-the-model). If you are using a known `probability_two_random_records_match` (and not estimating from `linker.estimate_probability_two_random_records_match`), then you will need to use `linker.estimate_u_using_random_sampling`, to find the u-values, and run `linker.estimate_parameters_using_expectation_maximisation` on a couple of different training blocking rules to estimate the m-values.

If that does not help/is not relevant, there are some ways to investigate what is occurring directly. If your data set is not too large you can call `linker.predict` without `threshold_match_probability` or `threshold_match_weight` (you only need to specify one of these btw - the one corresponding to a higher threshold will be used) - this will return predictions for _every_ record pair generated by your blocking rules. You can then manually query this predictions frame to see how these particular records are being scored. If that is not feasible you can also use `linker.compare_two_records(r1, r2)` passing the records as dicts to see the scores - see [here](https://moj-analytical-services.github.io/splink/demos/example_real_time_record_linkage.html#step-comparing-two-records). In either case you can also pass this scored record pair to `linker.waterfall_chart` to investigate why they are receiving the score they are, which may point you in the direction of which parameters are causing issues. `linker.match_weights_chart()` is probably useful here as well, to see if the parameters fit what you would expect from your data.

**Reply by reyvasquez-vh** [reply]
Thanks for the reply.  I forgot to copy/paste `estimate_probability_two_random_records_match`() and `estimate_u_using_random_sampling()` I added them above--not sure if they are both necessary.  In the end the print out says my model is fully trained.

I will try your recommendations above.  Thanks for taking the time to explain things in details.  I am running pyspark on AWS EMR.  I haven't had luck getting charting to work, even on EMR studio/notebook since moving away from DuckDB.
My dataset is 150Million records, but I will try` linker.predict()` without threshold and see how that goes.

Thanks again.

**Reply by RobinL** [reply]
The easiest way to do charts on remote servers is as follows:

```
from splink.charts import save_offline_chart
c = linker.match_weights_chart()
save_offline_chart(c.spec, "test_chart.html")
```

The chart now exists on disk as test_chart.html, and you then need to e.g. upload it into s3, so you can then download it onto your local machine to view.
